public/app/features/plugins/extensions/usePluginExtensions.tsx

```diff
@@ -1,64 +1,39 @@
 
 import { getPluginExtensions } from './getPluginExtensions';
-import { ReactivePluginExtensionsRegistry } from './reactivePluginExtensionRegistry';
-import { AddedComponentsRegistry } from './registry/AddedComponentsRegistry';
+import { PluginExtensionRegistries } from './registry/types';
 
-export function createUsePluginExtensions(
-  extensionsRegistry: ReactivePluginExtensionsRegistry,
-  addedComponentsRegistry: AddedComponentsRegistry
-) {
-  const observableRegistry = extensionsRegistry.asObservable();
-  const observableAddedComponentRegistry = addedComponentsRegistry.asObservable();
-  const cache: {
-    id: string;
-    extensions: Record<string, { context: GetPluginExtensionsOptions['context']; extensions: PluginExtension[] }>;
-  } = {
-    id: '',
-    extensions: {},
-  };
+export function createUsePluginExtensions(registries: PluginExtensionRegistries) {
+  const observableAddedComponentsRegistry = registries.addedComponentsRegistry.asObservable();
+  const observableAddedLinksRegistry = registries.addedLinksRegistry.asObservable();
 
   return function usePluginExtensions(options: GetPluginExtensionsOptions): UsePluginExtensionsResult<PluginExtension> {
-    const registry = useObservable(observableRegistry);
-    const addedComponentsRegistry = useObservable(observableAddedComponentRegistry);
-
-    if (!registry || !addedComponentsRegistry) {
-      return { extensions: [], isLoading: false };
-    }
-
-    if (registry.id !== cache.id) {
-      cache.id = registry.id;
-      cache.extensions = {};
-    }
-
-    // `getPluginExtensions` will return a new array of objects even if it is called with the same options, as it always constructing a frozen objects.
-    // Due to this we are caching the result of `getPluginExtensions` to avoid unnecessary re-renders for components that are using this hook.
-    // (NOTE: we are only checking referential equality of `context` object, so it is important to not mutate the object passed to this hook.)
-    const key = `${options.extensionPointId}-${options.limitPerPlugin}`;
-    if (cache.extensions[key] && cache.extensions[key].context === options.context) {
-      return {
-        extensions: cache.extensions[key].extensions,
-        isLoading: false,
-      };
-    }
-
-    const { extensions } = getPluginExtensions({
-      ...options,
-      registry,
+    const addedComponentsRegistry = useObservable(observableAddedComponentsRegistry);
+    const addedLinksRegistry = useObservable(observableAddedLinksRegistry);
+
+    const { extensions } = useMemo(() => {
+      if (!addedLinksRegistry || !addedComponentsRegistry) {
+        return { extensions: [], isLoading: false };
+      }
+
+      return getPluginExtensions({
+        extensionPointId: options.extensionPointId,
+        context: options.context,
+        limitPerPlugin: options.limitPerPlugin,
+        registryStates: { addedComponentsRegistry, addedLinksRegistry },
```

Comment:
**Nit:** just an idea - maybe we could simplify the callsite and the function declaration a bit if we would "flatten" the argument object:

```ts
return getPluginExtensions({
        extensionPointId: options.extensionPointId,
        context: options.context,
        limitPerPlugin: options.limitPerPlugin,
        addedComponents
        addedLinks,
```